simulation/agent/LineageAgent.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (C) 2021 Vaticle
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package com.vaticle.typedb.benchmark.simulation.agent;
+
+import com.vaticle.typedb.benchmark.common.concept.Country;
+import com.vaticle.typedb.benchmark.common.params.Context;
+import com.vaticle.typedb.benchmark.common.seed.RandomSource;
+import com.vaticle.typedb.benchmark.simulation.driver.Client;
+import com.vaticle.typedb.benchmark.simulation.driver.Session;
+import com.vaticle.typedb.benchmark.simulation.driver.Transaction;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class LineageAgent<TX extends Transaction> extends Agent<Country, TX> {
+
+    protected LineageAgent(Client<?, TX> client, Context context) {
+        super(client, context);
+    }
+
+    @Override
+    protected Class<? extends Agent> agentClass() {
+        return LineageAgent.class;
+    }
+
+    @Override
+    protected List<Country> regions() {
+        return context.seedData().countries();
+    }
+
+    @Override
+    protected List<Report> run(Session<TX> session, Country country, RandomSource random) {
+        List<Report> reports = new ArrayList<>();
+        try (TX tx = session.reasoningTransaction()) {
+            matchLineages(tx, country, context.startDay(), context.today());
+            if (context.isReporting()) {
+                throw new RuntimeException("Reports are not comparable for reasoning agents.");
+            }
+        }
+        return reports;
+    }
+
+    protected abstract void matchLineages(TX tx, Country country, LocalDateTime startDay, LocalDateTime today);
```

Comment:
Lineages is transitive so will find an enormous number of relations. These dates scope the amount we are looking for by anchoring that the ancestors are the people from the first year of the simulation and the descendants were born in the current year. Perhaps though we would rather bind only one end of this transitive query since that's a common use-case.